File: android/chesslibkt/src/main/java/ru/maksimbulva/chesslibkt/move/MoveBuilder.kt
package ru.maksimbulva.chesslibkt.move

import ru.maksimbulva.chesslibkt.Piece
import ru.maksimbulva.chesslibkt.board.Board
import ru.maksimbulva.chesslibkt.board.Square
import ru.maksimbulva.chesslibkt.encoded

inline class MoveBuilder(private val encoded: Int) {

    private val origin: Square
        get() = Square(encoded and SQUARE_MASK)

    private val dest: Square
        get() = Square((encoded shr 6) and SQUARE_MASK)

    constructor(piece: Piece, origin: Square)
        : this(origin.encoded or (piece.encoded shl 12))

    fun build() = Move(encoded).also { require(!it.isPromotion) }

    fun setDestSquare(dest: Square): MoveBuilder {
        return MoveBuilder(encoded or ((dest.encoded shl 6)))
    }

    fun setPromoteToPieceType(promoteTo: Piece): MoveBuilder {
        return MoveBuilder(encoded or Move.Promotion or (promoteTo.encoded shl 15))
    }

    fun setCapture(board: Board): MoveBuilder {
        val capturedPiece = board.getPieceTypeAt(dest)!!
        return MoveBuilder(encoded or Move.Capture or (capturedPiece.encoded shl 18))
    }

    fun setEnPassantCapture(): MoveBuilder {
        return MoveBuilder(encoded or (Move.Capture or Move.EnPassantCapture) or
                (Piece.Pawn.encoded shl 18))
    }

    fun setPawnDoubleMove(): MoveBuilder {
        return MoveBuilder(encoded or Move.PawnDoubleMove)
    }

    fun setLongCastle(): MoveBuilder {
        return MoveBuilder(encoded or Move.LongCastle)
    }

    fun setShortCastle(): MoveBuilder {
        return MoveBuilder(encoded or Move.ShortCastle)
    }
}

File: ChessApp/ChessEngine/Move/MoveBuilder.cs
﻿using ChessEngine.Board;

namespace ChessEngine.Move
{
    // Data are packed in a 32-bit integer for sake of memory efficiency
    internal readonly struct MoveBuilder
    {
        private readonly int packedValue;

        public MoveBuilder(Piece pieceToMove, BoardSquare originSquare)
        {
            packedValue = originSquare.IntValue;
            packedValue |= (int)pieceToMove << 12;
        }

        private MoveBuilder(int packedValue)
        {
            this.packedValue = packedValue;
        }

        public Move Build() => new Move(packedValue);

        public MoveBuilder SetDestSquare(BoardSquare destSquare)
        {
            return new MoveBuilder(packedValue | (destSquare.IntValue << 6));
        }

        public MoveBuilder SetCapture(Piece capturedPiece)
        {
            return new MoveBuilder(packedValue | Move.CaptureFlag | ((int)capturedPiece << 18));
        }

        public MoveBuilder SetEnPassantCapture()
        {
            return new MoveBuilder(packedValue | Move.CaptureFlag | Move.EnPassantCapture |
                ((int)Piece.Pawn << 18));
        }

        public MoveBuilder SetPromotion(Piece promoteToPiece)
        {
            return new MoveBuilder(packedValue | Move.Promotion | ((int)promoteToPiece << 15));
        }

        public MoveBuilder SetShortCastle()
        {
            return new MoveBuilder(packedValue | Move.ShortCastle);
        }

        public MoveBuilder SetLongCastle()
        {
            return new MoveBuilder(packedValue | Move.LongCastle);
        }

        public MoveBuilder SetPawnDoubleMove()
        {
            return new MoveBuilder(packedValue | Move.PawnDoubleMove);
        }
    }
}
